lib/sqlalchemy/dialects/sqlite/base.py

```diff
@@ -402,6 +402,253 @@ def set_sqlite_pragma(dbapi_connection, connection_record):
+    those specified using :paramref:`_schema.Column.onupdate`.
+    These values will not be exercised for an ON CONFLICT style of UPDATE,
+    unless they are manually specified in the
+    :paramref:`.Insert.on_conflict_do_update.set_` dictionary.
+
+In order to refer to the proposed insertion row, the special alias
+:attr:`~.sqlite.Insert.excluded` is available as an attribute on
+the :class:`_sqlite.Insert` object; this object creates an "excluded." prefix
+on a column, that informs the DO UPDATE to update the row with the value that
+would have been inserted had the constraint not failed::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(
+        id='some_id',
+        data='inserted value',
+        author='jlh')
+    do_update_stmt = stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value', author=stmt.excluded.author)
+        )
+    conn.execute(do_update_stmt)
+
+The :meth:`_expression.Insert.on_conflict_do_update` method also accepts
+a WHERE clause using the :paramref:`.Insert.on_conflict_do_update.where`
+parameter, which will limit those rows which receive an UPDATE::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(
+        id='some_id',
+        data='inserted value',
+        author='jlh')
+    on_update_stmt = stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value', author=stmt.excluded.author)
+        where=(my_table.c.status == 2)
+        )
+    conn.execute(on_update_stmt)
+
+``ON CONFLICT`` may also be used to skip inserting a row entirely
+if any conflict with a unique constraint occurs; below
+this is illustrated using the
+:meth:`~.sqlite.Insert.on_conflict_do_nothing` method::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(id='some_id', data='inserted value')
+    stmt = stmt.on_conflict_do_nothing(index_elements=['id'])
+    conn.execute(stmt)
+
+If ``DO NOTHING`` is used without specifying any columns or constraint,
+it has the effect of skipping the INSERT for any unique violation which
+occurs::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    stmt = insert(my_table).values(id='some_id', data='inserted value')
+    stmt = stmt.on_conflict_do_nothing()
+    conn.execute(stmt)
+
+The SQLite ``ON CONFLICT`` also supports multiple conflict targets. For example
+a unique index can be created over multiple columns, then this constraint will
+be inferred by listing the columns in index_elements::
+
+    unique_index = schema.Index(
+        "unique_index_name",
+        my_table.c.id,
+        my_table.c.data,
+        unique=True,
+    )
+
+    insert_stmt = insert(my_table).values(id='some_id', data='old_value')
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=['id', 'data'],
+        set_=dict(data='new_value')
+    )
```

Comment:
**mike bayer** (zzzeek) wrote:

this seems like more discussion about "index_elements", we can just refer to multiple columns when we talk about index_elements way up above, right?   folks will usually figure things out like this because we are close to the SQL syntax they want to use anyway.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2230